### ISSUE TITLE
Research: Handling of broken links in VA.gov build

### ISSUE THREAD

**Issue by mchelen-gov** [department-of-veterans-affairs/va.gov-team#19215]
### Build Pipeline

- Builds HTML assets
- Runs broken link check script
- Broken links are output to CSV file
- ~~Build fails if broken link CSV file exists~~
- [ ] Build fails if any broken links in CSV file are on pages not matched by ALLOWLIST
- Build failure caused by broken links notifies CMS team through Slack 
- [ ] Successful build with broken links on pages matching ALLOWLIST notifies CMS team through Slack with different priority level
- [ ] Successful build with broken links on pages matching ALLOWLIST publishes broken link CSV to public site
- CMS team triages broken links by fixing or contacting content editor
- [ ] CMS team triages broken links by contacting product team to update ALLOWLIST

Checkboxes represent changes and implementation tasks.

Note: ALLOWLIST refers to pages or patterns of va.gov URLs that are allowed to contain broken links. For example,
```
www.va.gov/my-favorite-pages/*
```
would ignore all broken links on `www.va.gov/my-favorite-pages/index.html` and `www.va.gov/my-favorite-pages/example/index.html`

[Mural Diagram](https://app.mural.co/t/departmentofveteransaffairs9999/m/departmentofveteransaffairs9999/1611968439525/a81486d1582f55c91c5f6ad7b5acd70b8c9b6629)

## Existing Process

link checker in FE build
- builds all html
- does not check external links
- if it finds broken link, creates csv file, csv blocks build
- cms team triages, either fixes themselves or contact content owner

## risks

- An influx of new editors in February - April will generate an influx of broken link incidents in the content layer. Current CMS help desk capacity is insufficient to handle growth in broken link incident triage in addition to regular support request from a projected user mass (~300). Mitigation strategy needs to be considered before CMS hits critical user mass.

### vets-website frontend
- react application
- header, footer, menus

### cms content editors
content editor can bypass [broken link checker](https://www.drupal.org/project/node_link_report)
- some links blocked by VA firewall
- some links might point to things that have not yet been published

### automated cms content import

**Comment by oksana-c** [comment]
## Notes 1/29/21

Participants: @mchelen @indytechcook @swirtSJW @daddison48 @oksana-c 

### Questions to review

* what types of broken links are responsibility of the CMS to address?
* how will broken link incidents reflect on CMS team capacity?
* how do we want to separate concerns w/ FE-tool vs CMS?
* what is the business ask re: broken links?


### Notes

* should content build  distinguish content layer broken links vs other broken links?
   * should the content build be blocked by broken links in content layer? consensus is - no
* what runs in content build vs full build?
  * broken links check runs at the end of build content stage in both content builds and full builds
  * if content build fails > full build fails too
* the load on CMS capacity will greatly increase once more editors are onboarded
* what type of commitment should we formulate if we do not block content release on broken links in content layer?
   * should CMS commit to fixing broken links in X hours?


#### Scenarios for broken links on content layer

1. page linked to is unpublished
2. faulty link in wysiwyg
3. link to a pdf that's on va.gov where somebody removed a file from managed files or upstream
4. URL to a linked node has changed. e.g. page moved under different menu, etc. These are not caught by node link checker - because Drupal can follow redirects stored within the CMS


### Next steps:
  * short-term:
    * generate public CSV w/ broken links
    * break the build only if broken links in content layer reach a certain threshold. _What threshold is acceptable for business?_
    * alerts will be posted in cms-team channel for CMS team to react
    * we need to establish broken links SLA. It's unreasonable to think that we'll be able to jump on every broken link in real time
       * there's 1 FTE on CMS help desk
       * majority of links would be passed on to authors with a message thru help desk
 * long-term:
    * offload broken links resolution to editors and their teams via near real-time notifications


### Actions items

1. Remove content build dependency on broken links until broken links threshold is reached.
   1. Establish threshold that is acceptable to the business.
1. Establish broken links SLA within CMS help desk processes.
1. Establish a process of triage where most links are passed on to the author and we shepherd/train them to fix content issues
1. Commit CSV that being currently generated for each build as a publicly accessible file.
1. Start planning for iteration to automate broken link triage process to bypass CMS help desk and send notifications to authors.

**Comment by oksana-c** [comment]
documenting proposal from @davidconlon 

> Platform has the broken link build stop so that a Veteran has a good experience on VA.gov and doesn't experience a broken link. I don't think that we should comprehensively change this without purposeful thought.
So here is an alternative suggestion: I suggest, instead, we create a whitelist of URLs/URL patterns where broken links are deemed acceptable by platform and product owners. (therefore thought and justification).
The change in the build process would be that instead of the suggestion of "broken link threshold" the build process checks this list to see if there is a broken link not on a URL/URL Pattern listed on the whitelist.  If there is a broken link on a non-whitelisted page, build stops. Else, Build proceeds. Either way the CMS team is notified if the build is stopped or the build is completed with known errors. (That way there will be a DIFF file to refer back to for successful builds to weed out)
> An example of this whitelist pattern would be:
https://www.va.gov/erie-health-care/*

link to Mural outlining proposed process https://app.mural.co/t/departmentofveteransaffairs9999/m/departmentofveteransaffairs9999/1611968439525/a81486d1582f55c91c5f6ad7b5acd70b8c9b6629

**Comment by davidconlon** [comment]
@mchelen I think this accurately captures what I proposed.

**Comment by johnhashva** [comment]
@davidconlon + @mchelen (and team) Overall, I like this "smarter/more targeted" approach.   Just to make sure I fully grok implications:

@davidconlon -- as use cases is this the right way to apply the thinking here: 

- 1 broken link on VA.gov HP = stop the build.
- 1 broken link on any Tier 1 Benefits Hub pages = stop the build (note: we could maybe refine this a bit)
- 1 or more broken links on About VA pages = does **NOT** stop the build (whitelist)

**Comment by swirtSJW** [comment]
In general this would be a big improvement over what we have now.  For long term stability it may be easier to list the critical pages that should stop the build.
Example:

- Home page
- all level 1 pages (va.gov/level-1/level-2/level-3 )
- ?

Might make it easier for debugging.

**Comment by johnhashva** [comment]
Just received additional context from @davidconlon and posting here for sharing since it may be helpful to others: 

- Right now, VA.gov build stops on any broken links.

- What we're proposing is an "escape hatch" so that specific URLs/URL patterns don't block the build if they have a broken link on it.

- For example: VAMC website Erie-health-care in dual state (low impact to veteran) may have MANY broken links as it's being worked on/cleaned. Because of the low-impact to veterans and a non-OCTO editor, I would request FE-tools add the erie-health-care pattern to the whitelist so that erie doesn't block the rest of VA.gov from building.
However, if a broken link is found on a non-whitelisted URL, then the current process continues.

**Comment by mchelen-gov** [comment]
@johnhashva That's right, great summary! CMS team would still be notified about `Erie-health-care` broken links but they would not block the build. Then the content owner would need to work with CMS team to get those resolved eventually so the entry can be removed.

**Comment by mchelen-gov** [comment]
> In general this would be a big improvement over what we have now. For long term stability it may be easier to list the critical pages that should stop the content release.
> Example:
> 
> * Home page
> * all level 1 pages (va.gov/level-1/level-2/level-3 )
> * ?
> Might make it easier for debugging as well as more scalable.
> 

@swirtSJW That's an interesting approach, I wonder how to decide what the "critical" page list would include, such as would homepage + level 1 be sufficient?

> In addition there should still be a count threshold. More than 75 broken links ought to break the content release.

@swirtSJW That's a great point, because this approach handles the situation @johnhashva described of a new VAMC going live, but does not address one-off broken links created in unrelated pages.

**Comment by oksana-c** [comment]
On the topic of CMS help desk SLA:

if the build can send two different Slack messages (warning and error) to #cms-team channel, then we may need two CSV files generated:
    * (1) CSV that contains URLs based on ALLOWLIST (file URL available in Slack ⚠️ Warning message)
    * (2) CSV that contains all other URLs (file URL available in Slack :x: Error message)

If we can support this differentiation, then the CMS help desk will be able to distinguish between two levels of urgency for fixing the links and have a better troubleshooting experience overall.
thoughts?

/cc @mchelen @meganhkelley @davidconlon @johnhashva @indytechcook @swirtSJW @daddison48

**Comment by swirtSJW** [comment]
I don't think we should do two separate CSVs.   Either way we need to repair the broken links and having to look to multiple sources in multiple locations makes this more complicated.
We will already have to look at two different locations.   
1. Content release completes, the CSV will be public within the release.
2. Content release fails, the CSV will be in jenkiins.

**Comment by mchelen-gov** [comment]
Ok it sounds like our implementation options are:
1. ALLOWLIST approach
   - Sections of website can be set to allow broken links as part of deployment lifecycle
2. Threshold approach
   - One-off broken links will not block entire build from deploying
3. Both

@davidconlon @johnhashva Does combining these approaches make sense to enable rollout and keep content publishing fast?

@meganhkelley Can someone from FE team provide feedback on implementation of these options?

**Comment by mchelen-gov** [comment]
From discussion with CMS team and @swirtSJW it sounds like:
1. Broken link risk from VAMC rollout is low because of close coordination during publication process
2. Broken link risk from one-off links is high as number of editors increase

Based on this the implementation suggestion is:
1. Do not use rolling `allowlist` for VAMC sites
2. Use a threshold so that 10 or less broken links do not block build
3. Top level `va.gov` should not allow any broken links
4. CMS team should be alerted when there are broken links, even if they do not break the build

**Comment by jefflbrauer** [comment]
@mchelen Per slack chat, the team continues to support the above recommendation, with the one clarification on #4.  The broken link alerts should be in the same channel as content release failure notifications (currently #cms-notifications), and a broken link alert must distinguish whether or not it has broken the build. According to our criteria, this would happen if the broken link was either: 

1. on top level va.gov
2. the 11th broken link in the same build

**Comment by acrollet** [comment]
CMS team discussed implementing the approach in https://github.com/department-of-veterans-affairs/va.gov-team/issues/19215#issuecomment-809496188 today with @mchelen-gov. I've created https://github.com/department-of-veterans-affairs/va.gov-cms/issues/5050 in the core team backlog to implement this approach.

**Comment by mchelen-gov** [comment]
closing in favor of https://github.com/department-of-veterans-affairs/va.gov-cms/issues/5050
